### PR TITLE
Reset routing if redirect fails.

### PR DESCRIPTION
**Pull request by nihohit**
*Issue #, if available:*
https://github.com/aws/glide-for-redis/issues/741

*Description of changes:*
Based on https://github.com/redis-rs/redis-rs/pull/968, which fixes issues in redis-rs.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: redis/src/cluster_async/mod.rs [hunk 1]
```diff
@@ -377,17 +380,67 @@ struct RequestInfo<C> {
...
-                CmdArg::Cmd { routing, .. } => {
-                    *routing = redirect.into();
-                }
+                CmdArg::Cmd { routing, .. } => match routing {
+                    InternalRoutingInfo::SingleNode(route) => {
+                        let redirect = InternalSingleNodeRouting::Redirect {
+                            redirect,
+                            previous_routing: Box::new(std::mem::replace(
+                                route,
+                                InternalSingleNodeRouting::Random,
```
Comment (barshaul) [inline redis/src/cluster_async/mod.rs:390]: I don't understand why do you replace 'route' with Random node, when is it going to be used?
please add a comment explaining the logic

File: redis/src/cluster_async/mod.rs [hunk 2]
```diff
@@ -377,17 +380,67 @@ struct RequestInfo<C> {
...
+                    InternalRoutingInfo::MultiNode(_) => {
+                        panic!("Cannot redirect multinode requests")
+                    }
+                },
                 CmdArg::Pipeline { route, .. } => {
+                    let redirect = InternalSingleNodeRouting::Redirect {
+                        redirect,
+                        previous_routing: Box::new(std::mem::replace(
+                            route,
+                            InternalSingleNodeRouting::Random,
```
Comment (barshaul) [inline redis/src/cluster_async/mod.rs:405]: same

File: redis/src/cluster_async/mod.rs [hunk 3]
```diff
@@ -377,17 +380,67 @@ struct RequestInfo<C> {
...
+
+    fn reset_redirect(&mut self) {
+        match &mut self.cmd {
+            CmdArg::Cmd { routing, .. } => {
+                if let InternalRoutingInfo::SingleNode(InternalSingleNodeRouting::Redirect {
+                    previous_routing,
+                    ..
+                }) = routing
+                {
+                    let mut previous_routing = std::mem::replace(
```
Comment (barshaul) [inline redis/src/cluster_async/mod.rs:422]: still can't understand :( the var names are confusing too.
you take previous_routing_1 and put Random in it, then previous_routing_2 takes the old value of previous_routing_1.
so now
previous_routing_2 = previous_routing_1
previous_routing_1 = Random
and then you swap routing, which contains a reference to previous_routing_1, and replace it with previous_routing_2?
help.

File: redis/src/cluster_async/mod.rs (hunk 1 above)
Comment (shachlanAmazon) [reply redis/src/cluster_async/mod.rs:390]: https://doc.rust-lang.org/std/mem/fn.replace.html

**Review by barshaul — APPROVED** [review]

**Pull request merged**